File: DSA_Imp/Arrays/SpiralMatrix.java
// Give a matrix N*M. print all the elements of matrix in spiral order.
// input
// {{1,2,3,4},
//  {5,6,7,8},
//  {9,10,11,12},
//  {13,14,15,16}}
// Output = {1,2,3,4,8,12,16,15,14,13,9,5,6,7,11,10}
// Time complexity = O(n*m) 
// Space complexity = O(1)

package DSA_Imp.Arrays;

public class SpiralMatrix {
    public static void main(String[] args){
        int[][] arr ={{1,2,3,4,5},
                      {6,7,8,9,10},
                      {11,12,13,14,15}};
        matrix(arr);
    }
    public static void matrix(int[][] arr){
        int n = arr.length;
        int m = arr[0].length;
        int rmin = 0, rmax=arr.length-1;
        int cmin = 0, cmax= arr[0].length-1;
        int count =0;
        while(count < n * m){
            // top boundary
            for(int col = cmin; col <= cmax && count< n*m; col++){
                System.out.print(arr[rmin][col] + " ");
                count++;
            }
            rmin++;

            // rignt boundary
            for(int row = rmin; row <= rmax && count< n*m; row++){
                System.out.print(arr[row][cmax] + " ");
                count++;
            }
            cmax--;

            // bottom boundary
            for(int col = cmax; col >= cmin && count< n*m; col--){
                System.out.print(arr[rmax][col] + " ");
                count++;
            }
            rmax--;

            // left boundary
            for(int row = rmax; row >= rmin && count< n*m; row--){
                System.out.print(arr[row][cmin] + " ");
                count++;
            }
            cmin++;
        } 
    }
}

File: Matrix/SpiralMatrix.py
# Input: matrix = [[1,2,3],[4,5,6],[7,8,9]]
# Output: [1,2,3,6,9,8,7,4,5]

from typing import List

class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        result = []
        if not matrix or not matrix[0]:
            return result
        
        top, bottom = 0, len(matrix) - 1
        left, right = 0, len(matrix[0]) - 1
        
        while top <= bottom and left <= right:
            # Traverse from left to right across the top row
            for i in range(left, right + 1):
                result.append(matrix[top][i])
            top += 1  # Move the top boundary down
            
            # Traverse from top to bottom along the right column
            for i in range(top, bottom + 1):
                result.append(matrix[i][right])
            right -= 1  # Move the right boundary left
            
            if top <= bottom:
                # Traverse from right to left across the bottom row
                for i in range(right, left - 1, -1):
                    result.append(matrix[bottom][i])
                bottom -= 1  # Move the bottom boundary up
            
            if left <= right:
                # Traverse from bottom to top along the left column
                for i in range(bottom, top - 1, -1):
                    result.append(matrix[i][left])
                left += 1  # Move the left boundary right
        
        return result